### PR TITLE
fix for NRE at StationComponent.InternalTickRemote

### LINKED ISSUE #59: NullReferenceException after joining the game with planetary stations
This exception is thrown every FixedUpdate() call to the player who has joined the game. Seems to be caused by the stations.
Error Message

```NullReferenceException: Object reference not set to an instance of an object
  at StationComponent.InternalTickRemote (Int32 timeGene, Double dt, Single shipSailSpeed, Single shipWarpSpeed, Int32 shipCarries, .StationComponent[] gStationPool, .AstroPose[] astroPoses, VectorLF3 relativePos, Quaternion relativeRot, Boolean starmap, System.Int32[] consumeRegister) [0x00000] in <filename unknown>:0 
  at PlanetTransport.GameTick (Int64 time, Boolean isActive) [0x00000] in <filename unknown>:0 
  at PlanetFactory.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameData.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameMain.FixedUpdate () [0x00000] in <filename unknown>:0
```

Steps to reproduce:

1. You can use mine [savegame](https://drive.google.com/file/d/1CTA6JhYSiIVhXN6szeprketV5ieXdcDm/view?usp=sharing)
2. Launch the game normally as a single-player and from [Esc] game menu host the game.
3. Join the game with the second player and observe the error.

### PR DESCRIPTION
**Pull request by sp00ktober**
This fixes #59 by requesting the needed `FactoryData` of Planets that contain ILS/PLS stations when joining the game.

### CODE REVIEW

**Review by Mushroom — CHANGES_REQUESTED** [review]
As discussed in discord, I disagree with this approach - we should not be loading in all the factories, we should just get authoritative updates on logistics vessel positions/travels from the host

**Comment by sp00ktober** [comment]
This will be a bigger PR again i guess. This is just because the fix for #59 turns out to be a working sync implementation for ILS and its ships.

- [x] Fix issue with prebuildings (buildings are sometimes unclickable after placed, prebuildings do only vanish after every building is placed)
- [ ] Sync ILS UI
- [ ] Sync item transport done by ships
- [ ] Sync ship rendering

File: NebulaPatcher/Patches/Dynamic/GalacticTransport_Patch.cs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+                    Debug.Log(i + ": null");
+                }
+            }
+        }
+
+        [HarmonyPostfix]
+        [HarmonyPatch("RemoveStationComponent")]
+        public static void RemoveStationComponent_Postfix(GalacticTransport __instance, int gid)
+        {
+            if (__instance.stationRecycleCursor > 0)
```
Comment (Therzok) [inline NebulaPatcher/Patches/Dynamic/GalacticTransport_Patch.cs:101]: Shouldn't this be a message sent to the host to try and remove the station component?

Especially, since this should be reference counted per player, so you don't remove station components that are still in use?

File: NebulaPatcher/Patches/Dynamic/GalacticTransport_Patch.cs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+                {
+                    __instance.stationRecycleCursor = 0;
+                }
+            }
+            int debug = 0;
+            if (__instance.stationRecycleCursor > 0)
+            {
+                Debug.Log((__instance.stationPool[__instance.stationRecycle[__instance.stationRecycleCursor - 1]] == null) ? "ITS NULL" : "ITS NOT NULL");
+                Debug.Log("recycle " + __instance.stationRecycleCursor);
+                debug = __instance.stationRecycle[__instance.stationRecycleCursor-1];
```
Comment (Therzok) [inline NebulaPatcher/Patches/Dynamic/GalacticTransport_Patch.cs:46]: Uhm, you shouldn't do -1 here. `stationRecycle[0]` is _never_ used. Only indices [1..capacity]

File: NebulaWorld/Logistics/StationUIManager.cs [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+using NebulaModel.DataStructures;
+using NebulaModel.Networking;
+using UnityEngine;
+using UnityEngine.EventSystems;
+
+namespace NebulaWorld.Logistics
+{
+    public static class StationUIManager
+    {
+        private static ThreadSafeDictionary<int, List<NebulaConnection>> StationUISubscribers;
```
Comment (Therzok) [inline NebulaWorld/Logistics/StationUIManager.cs:14]: This is gone in latest master.

File: NebulaPatcher/Patches/Dynamic/GalacticTransport_Patch.cs (hunk 2 above)
Comment (sp00ktober) [reply NebulaPatcher/Patches/Dynamic/GalacticTransport_Patch.cs:55]: decrement
it does access index 0 too :)

File: NebulaPatcher/Patches/Dynamic/GalacticTransport_Patch.cs (hunk 1 above)
Comment (sp00ktober) [reply NebulaPatcher/Patches/Dynamic/GalacticTransport_Patch.cs:111]: yes i probably should make this a message to the host too, thanks.

File: NebulaClient/PacketProcessors/Logistics/ILSAddStationComponentResponseProcessor.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    [RegisterPacketProcessor]
+    public class ILSAddStationComponentResponseProcessor: IPacketProcessor<ILSAddStationComponentResponse>
+    {
+        public void ProcessPacket(ILSAddStationComponentResponse packet, NebulaConnection conn)
+        {
+            for(int i = 0; i < ILSShipManager.AddStationComponentQueue[packet.planetId].Count; i++)
+            {
+                if(ILSShipManager.AddStationComponentQueue[packet.planetId][i].shipDockPos == DataStructureExtensions.ToVector3(packet.shipDockPos))
+                {
+                    LocalPlayer.PatchLocks["GalacticTransport"] = true;
```
Comment (Therzok) [inline NebulaClient/PacketProcessors/Logistics/ILSAddStationComponentResponseProcessor.cs:21]: There is a toggleswitch abstraction now that can be used for this.

File: NebulaModel/Packets/Logistics/ILSgStationPoolSync.cs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+                                    int[] shipIndex,
+                                    Double3[] shipPos,
+                                    Float4[] shipRot,
+                                    Float3[] shipVel,
+                                    float[] shipSpeed,
+                                    Float3[] shipAngularVel,
+                                    Double3[] shipPPosTemp,
+                                    Float4[] shipPRotTemp)
+        {
+            this.stationGId = new int[stationGId.Length];
```
Comment (Therzok) [inline NebulaModel/Packets/Logistics/ILSgStationPoolSync.cs:70]: These could be `Array.Clone()`. Something like `this.stationGId = (int[])stationGId.Clone()`.

File: NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    {
+        public void ProcessPacket(StationUIInitialSyncRequest packet, NebulaConnection conn)
+        {
+            StationComponent[] gStationPool = GameMain.data.galacticTransport.stationPool;
+            if(packet.stationGId < gStationPool.Length)
+            {
+                StationComponent stationComponent = gStationPool[packet.stationGId];
+                if(stationComponent != null)
+                {
+                    List<int> itemId = new List<int>();
```
Comment (Therzok) [inline NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs:21]: Is this sending data for entries that aren't set?

Comment (Therzok) [reply NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs:21]: Also, you can pre-allocate arrays here.

```csharp
var storage = storageComponent.storage;
var itemId = new int[storage.Length];
...

for (int i = 0; i < storage.Length; ++i) {
    var entry = storage[i];
    itemId[i] = entry.itemId;
    itemCountMax = entry.max;
    ..
}

File: NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+                    List<int> itemCountMax = new List<int>();
+                    List<int> localLogic = new List<int>();
+                    List<int> remoteLogic = new List<int>();
+                    for(int i = 0; i < stationComponent.storage.Length; i++)
+                    {
+                        itemId.Add(stationComponent.storage[i].itemId);
+                        itemCountMax.Add(stationComponent.storage[i].max);
+                        switch (stationComponent.storage[i].localLogic)
```
Comment (Therzok) [inline NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs:29]: You can simply use `(int)localLogic`

File: NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+                            case (ELogisticStorage.Supply):
+                                localLogic.Add(1);
+                                break;
+                            case (ELogisticStorage.Demand):
+                                localLogic.Add(2);
+                                break;
+                        }
+                        switch (stationComponent.storage[i].remoteLogic)
+                        {
+                            case (ELogisticStorage.None):
```
Comment (Therzok) [inline NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs:43]: Same here, `(int)remoteLogic`

File: NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+                    }
+                    StationUIInitialSync packet2 = new StationUIInitialSync(packet.stationGId,
+                                                                            stationComponent.tripRangeDrones,
+                                                                            stationComponent.tripRangeShips,
+                                                                            stationComponent.deliveryDrones,
+                                                                            stationComponent.deliveryShips,
+                                                                            stationComponent.warpEnableDist,
+                                                                            stationComponent.warperNecessary,
+                                                                            stationComponent.includeOrbitCollector,
+                                                                            itemId.ToArray(),
```
Comment (Therzok) [inline NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs:62]: If you pre-allocate the arrays above, you don't need to ToArray() here.

File: NebulaClient/PacketProcessors/Logistics/StationUIInitialSyncProcessor.cs [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+                    stationComponent.warperNecessary = packet.warperNecessary;
+                    stationComponent.includeOrbitCollector = packet.includeOrbitCollector;
+                    for(int i = 0; i < packet.itemId.Length; i++)
+                    {
+                        stationComponent.storage[i].itemId = packet.itemId[i];
+                        stationComponent.storage[i].max = packet.itemCountMax[i];
+                        switch (packet.localLogic[i])
+                        {
+                            case 0:
+                                stationComponent.storage[i].localLogic = ELogisticStorage.None;
```
Comment (Therzok) [inline NebulaClient/PacketProcessors/Logistics/StationUIInitialSyncProcessor.cs:42]: Get rid of switch :p `localLogic = (ELogisticStorage)intValue;`

File: NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs (hunk 6 above)
Comment (sp00ktober) [reply NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs:21]: Ah thanks, i should put the sending code inside the check.

File: NebulaWorld/Logistics/ILSShipManager.cs [hunk 11]
```diff
@@ -209,13 +218,64 @@ public static void AddTakeItem(ILSShipItems packet)
...
+                Debug.Log(GameMain.galaxy.PlanetById(stationComponent.planetId).displayName + ": dir: " + stationComponent.workShipDatas[packet.origShipIndex].direction + " carry " + stationComponent.workShipDatas[packet.origShipIndex].itemId + "(" + stationComponent.workShipDatas[packet.origShipIndex].itemCount + ")");
+            }
+        }
+
+        private static void RefreshValuesUI(StationComponent stationComponent, int storageIndex)
+        {
+            UIStationWindow stationWindow = UIRoot.instance.uiGame.stationWindow;
+            if (stationWindow != null && (int)AccessTools.Field(typeof(UIStationWindow), "_stationId")?.GetValue(stationWindow) == stationComponent.id)
+            {
+                UIStationStorage[] stationStorageUI = (UIStationStorage[])AccessTools.Field(typeof(UIStationWindow), "storageUIs").GetValue(stationWindow);
```
Comment (Therzok) [inline NebulaWorld/Logistics/ILSShipManager.cs:275]: please cache these 3 AccessTools into FieldRef delegates. From the looks of it, these are called often, and they create allocations - possibly introducing more points for host/client to deync.

**Comment by sp00ktober** [comment]
I leave it as it is now, the rest that is missing or thats still bugged will be covered by individual issues.
With this you should be able to play the game and make use of ILS/PLS and Orbital Collectors.
Both clients and host should be able to place the stations anywhere they like and configure them, add and remove items and objects (drones, ships, warpers).

File: NebulaPatcher/Patches/Dynamic/GalacticTransport_Patch.cs (hunk 1 above)
Comment (sp00ktober) [reply NebulaPatcher/Patches/Dynamic/GalacticTransport_Patch.cs]: This is removed in current code, i added an issue for the error popping up when a client removes an ILS (#241 )

File: NebulaWorld/Logistics/StationUIManager.cs (hunk 3 above)
Comment (sp00ktober) [reply NebulaWorld/Logistics/StationUIManager.cs]: Im using the normal `Dictionary` there for now

File: NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs (hunk 9 above)
Comment (sp00ktober) [reply NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs:81]: is it bad to use `ToArray()` in this situation?

File: NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs (hunk 7 above)
Comment (sp00ktober) [reply NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs:45]: changed in next commit

File: NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs (hunk 8 above)
Comment (sp00ktober) [reply NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs:59]: changed in next commit

File: NebulaClient/PacketProcessors/Logistics/StationUIInitialSyncProcessor.cs (hunk 10 above)
Comment (sp00ktober) [reply NebulaClient/PacketProcessors/Logistics/StationUIInitialSyncProcessor.cs:61]: changed in next commit :D

File: NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs (hunk 6 above)
Comment (sp00ktober) [reply NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs:33]: preallocation is done in next commit :)

File: NebulaModel/Packets/Logistics/ILSgStationPoolSync.cs (hunk 5 above)
Comment (sp00ktober) [reply NebulaModel/Packets/Logistics/ILSgStationPoolSync.cs:70]: implemented like that in next commit.

File: NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs (hunk 9 above)
Comment (Therzok) [reply NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs:81]: ToArray() creates a copy of the collection.

```csharp
var list = new List<int>(); // this uses an int[8] internally
for (int i = 0; i < source.Length; ++i) {
   // If the list array is full, it doubles it, so you get another allocation for
   // int[16], int[32], int[64], etc.
   // you also need to copy hte values to the new array
   list.Add(source[i]); 
}
var result = list.ToArray(); // another `array[list.Count]` allocation + copy
```

if you just do
```csharp
// just one allocation and copy
var result = (int[])source.Clone();

// What clone does
var result = new int[source.Length];
Array.Copy(source, result, source.Length);

// Similarly you have
var result = Array.ConvertAll(source, x => x);
```

Comment (sp00ktober) [reply NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs:81]: Thanks for the detail example, that makes it really clear 😄

File: NebulaWorld/Logistics/ILSShipManager.cs (hunk 11 above)
Comment (sp00ktober) [reply NebulaWorld/Logistics/ILSShipManager.cs:282]: included in the next commit.

File: NebulaModel/Packets/Logistics/ILSShipData.cs [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+        public int planetAStationGID { get; set; }
+        public int planetBStationGID { get; set; }
+        public int origShipIndex { get; set; }
+        public int warperCnt { get; set; }
+        public int stationWarperCnt { get; set; }
+
+        public ILSShipData() { }
+        public ILSShipData(bool IdleToWork, int planetA, int planetB, int itemId, int itemCount, int AGID, int BGID, int origShipIndex, int warperCnt, int stationWarperCnt)
+        {
+            this.IdleToWork = IdleToWork;
```
Comment (hubastard) [inline NebulaModel/Packets/Logistics/ILSShipData.cs:19]: `IdleToWork` should start with a lowerCase

File: NebulaModel/Packets/Logistics/ILSShipItems.cs [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+        public bool AddItem { get; set; }
+        public int itemId { get; set; }
+        public int itemCount { get; set; }
+        public int origShipIndex { get; set; }
+        public int stationGID { get; set; }
+
+        public ILSShipItems() { }
+        public ILSShipItems(bool AddItem, int itemId, int itemCount, int origShipIndex, int stationGID)
+        {
+            this.AddItem = AddItem;
```
Comment (hubastard) [inline NebulaModel/Packets/Logistics/ILSShipItems.cs:14]: `AddItem` should start with a lowercase

File: NebulaModel/Packets/Logistics/ILSSystemStorageSync.cs [hunk 14]
```diff
@@ -0,0 +1,8 @@
+﻿namespace NebulaModel.Packets.Logistics
+{
+    public class ILSSystemStorageSync
+    {
+        public int[] stationGId { get; set; }
+        public int[] storageStationGId { get; set; }
```
Comment (hubastard) [inline NebulaModel/Packets/Logistics/ILSSystemStorageSync.cs:6]: No constructor with array copy here ?

File: NebulaPatcher/NebulaPlugin.cs [hunk 15]
```diff
@@ -43,6 +44,14 @@ private void InitPatches()
...
             {
                 var assemblies = AppDomain.CurrentDomain.GetAssemblies();
+                //Environment.SetEnvironmentVariable("MONOMOD_DMD_TYPE", "mb");
+                //Environment.SetEnvironmentVariable("MONOMOD_DMD_DUMP", "./mmdump");
+                /*
+                foreach (FileInfo file in new DirectoryInfo("./mmdump").GetFiles())
+                {
+                    file.Delete();
+                }
+                */
```
Comment (hubastard) [inline NebulaPatcher/NebulaPlugin.cs:54]: Maybe we could add another build target that enables this

File: NebulaWorld/Logistics/StationUIManager.cs [hunk 16]
```diff
@@ -0,0 +1,464 @@
...
+                    UpdateStorageUI(packet);
+                }
+                else
+                {
+                    UpdateSettingsUI(packet);
+                }
+            }
+        }
+
+        private static void UpdateSettingsUIBackground(StationUI packet, PlanetData pData, int stationGId)
```
Comment (hubastard) [inline NebulaWorld/Logistics/StationUIManager.cs:105]: This method seems heavy and it looks like there is some code copied from the game. We should probably look if we can simplify it or call the actual game code instead.

File: NebulaModel/Packets/Logistics/ILSSystemStorageSync.cs (hunk 14 above)
Comment (sp00ktober) [reply NebulaModel/Packets/Logistics/ILSSystemStorageSync.cs:6]: Ah yes, i started to work on something but made an individual issue out of it. Im gonna remove this file for now.

File: NebulaPatcher/NebulaPlugin.cs (hunk 15 above)
Comment (sp00ktober) [reply NebulaPatcher/NebulaPlugin.cs:54]: Im not sure how this would be done :/

File: NebulaClient/PacketProcessors/Logistics/ILSAddStationComponentResponseProcessor.cs (hunk 4 above)
Comment (sp00ktober) [reply NebulaClient/PacketProcessors/Logistics/ILSAddStationComponentResponseProcessor.cs:21]: I *think* this is not the usecase here. What i need to do is set a boolean that i can check for in my Prefix patches to determine if they should run the patch code or the vanilla code in order to avoid endless loops.

File: NebulaWorld/Logistics/StationUIManager.cs (hunk 16 above)
Comment (sp00ktober) [reply NebulaWorld/Logistics/StationUIManager.cs:105]: I am calling the vanilla game functions in `UpdateSettingsUI()`, but since they belong to `UIStationWindow` i cant call them in `UpdateSettingsUIBackground()` since the window is closed at that time.

I could use reverse patching to get a modified version of the original method which i then could use there, but in the end it would be more complicated code including transpilers to get the same done.

I can see how i can make the method a bit easier to read.
basically its the functionality of 
grafik

**Review by Therzok — APPROVED** [review]

File: NebulaHost/PacketProcessors/Logistics/ILSgStationPoolSyncRequestProcessor.cs [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+        public void ProcessPacket(ILSRequestgStationPoolSync packet, NebulaConnection conn)
+        {
+            Player player = playerManager.GetPlayer(conn);
+            if(player == null)
+            {
+                player = playerManager.GetSyncingPlayer(conn);
+            }
+            if (player != null)
+            {
+                List<int> stationGId = new List<int>();
```
Comment (Therzok) [inline NebulaHost/PacketProcessors/Logistics/ILSgStationPoolSyncRequestProcessor.cs:27]: This is still ulsing lists and could be using arrays though.

Comment (sp00ktober) [reply NebulaHost/PacketProcessors/Logistics/ILSgStationPoolSyncRequestProcessor.cs:27]: ups

File: NebulaPatcher/NebulaPlugin.cs (hunk 15 above)
Comment (hubastard) [reply NebulaPatcher/NebulaPlugin.cs:54]: I'll look at doing this later

**Review by hubastard — APPROVED** [review]

File: NebulaHost/PacketProcessors/Logistics/ILSgStationPoolSyncRequestProcessor.cs [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+                int[] stationId = new int[countILS];
+                Float3[] DockPos = new Float3[countILS];
+                Float4[] DockRot = new Float4[countILS];
+                int[] planetId = new int[countILS];
+                int[] workShipCount = new int[countILS];
+                int[] idleShipCount = new int[countILS];
+                ulong[] workShipIndices = new ulong[countILS];
+                ulong[] idleShipIndices = new ulong[countILS];
+
+                int[] shipStationGId = new int[countILS * 10];
```
Comment (Mushroom) [inline NebulaHost/PacketProcessors/Logistics/ILSgStationPoolSyncRequestProcessor.cs:52]: I feel like a const as opposed to a fixed "10" would work better here, in the event someone wants compatibility with a mod that increases capacities, or the game has ILSes tweaked in the future - I see it repeated further down also

File: NebulaHost/PacketProcessors/Logistics/StationSubscribeUIUpdatesProcessor.cs [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+{
+    [RegisterPacketProcessor]
+    class StationSubscribeUIUpdatesProcessor: IPacketProcessor<StationSubscribeUIUpdates>
+    {
+        public void ProcessPacket(StationSubscribeUIUpdates packet, NebulaConnection conn)
+        {
+            if (packet.subscribe)
+            {
+                StationUIManager.AddSubscriber(packet.stationGId, conn);
+                // TODO: sync current state to client
```
Comment (Mushroom) [inline NebulaHost/PacketProcessors/Logistics/StationSubscribeUIUpdatesProcessor.cs:17]: Intentional TODO here?

File: NebulaModel/Networking/Serialization/NetPacketProcessor.cs [hunk 20]
```diff
@@ -22,8 +22,8 @@ private static class HashCache<T>
         private Queue<PendingPacket> pendingPackets = new Queue<PendingPacket>();
 
         public bool SimulateLatency = false;
-        public int SimulatedMinLatency = 20;
-        public int SimulatedMaxLatency = 50;
+        public int SimulatedMinLatency = 15;
```
Comment (Mushroom) [inline NebulaModel/Networking/Serialization/NetPacketProcessor.cs:25]: Should these values be changed? They are unrelated to ILSes

File: NebulaPatcher/NebulaPlugin.cs [hunk 21]
```diff
@@ -43,6 +44,14 @@ private void InitPatches()
             try
             {
                 var assemblies = AppDomain.CurrentDomain.GetAssemblies();
+                //Environment.SetEnvironmentVariable("MONOMOD_DMD_TYPE", "mb");
```
Comment (Mushroom) [inline NebulaPatcher/NebulaPlugin.cs:47]: Do not leave commented out code in the codebase

File: NebulaPatcher/Patches/Transpilers/PlayerAction_Mine_Patch.cs [hunk 22]
```diff
@@ -30,7 +30,7 @@ static IEnumerable<CodeInstruction> GameTick_Transpiler(IEnumerable<CodeInstruct
                     new CodeMatch(OpCodes.Ldarg_0),
                     new CodeMatch(OpCodes.Ldfld),
                     new CodeMatch(i => i.opcode == OpCodes.Callvirt && ((MethodInfo)i.operand).Name == "GetVegeData"))
-                .Insert(Transpilers.EmitDelegate<Func<int, int>>(miningId =>
+                .Insert(HarmonyLib.Transpilers.EmitDelegate<Func<int, int>>(miningId =>
```
Comment (Mushroom) [inline NebulaPatcher/Patches/Transpilers/PlayerAction_Mine_Patch.cs:33]: Why is this required, is there a namespace collision now?

File: NebulaWorld/Logistics/ILSShipManager.cs [hunk 23]
```diff
@@ -0,0 +1,304 @@
...
+                            break;
+                        }
+                    }
+                }
+                else
+                {
+                    int itemId = packet.itemId;
+                    int itemCount = packet.itemCount;
+                    stationComponent.TakeItem(ref itemId, ref itemCount);
+                    /*
```
Comment (Mushroom) [inline NebulaWorld/Logistics/ILSShipManager.cs:231]: Don't leave commented out code in

**Review by Mushroom — CHANGES_REQUESTED** [review]
A few specific comments where I thought they were needed.

I dislike the lack of comments in this pull request, someone unfamiliar with this codebase (or this part of it, in my case) will have absolutely no idea what si going on a lot of the time any why certain things are done.

I would like to see some more comments that explain the code.

**Review by Mushroom — APPROVED** [review]

File: NebulaClient/PacketProcessors/Logistics/ILSAddStationComponentResponseProcessor.cs [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+namespace NebulaClient.PacketProcessors.Logistics
+{
+    [RegisterPacketProcessor]
+    public class ILSAddStationComponentResponseProcessor: IPacketProcessor<ILSAddStationComponentResponse>
+    {
+        public void ProcessPacket(ILSAddStationComponentResponse packet, NebulaConnection conn)
+        {
+            for(int i = 0; i < ILSShipManager.AddStationComponentQueue[packet.planetId].Count; i++)
+            {
+                if(ILSShipManager.AddStationComponentQueue[packet.planetId][i].shipDockPos == DataStructureExtensions.ToVector3(packet.shipDockPos))
```
Comment (Baldie-dev) [inline NebulaClient/PacketProcessors/Logistics/ILSAddStationComponentResponseProcessor.cs:23]: Is it possible to put `DataStructureExtensions.ToVector3(packet.shipDockPos)` before the `for` loop? To prevent allocation of `new Vector3()` inside the loop.

File: NebulaHost/PacketProcessors/Logistics/ILSgStationPoolSyncRequestProcessor.cs [hunk 25]
```diff
@@ -0,0 +1,155 @@
...
+ * Whenever a client connects we sync the current state of all ILS and ships to them
+ * resulting in a quite large packet but its only sent one time upon client connect.
+ */
+namespace NebulaHost.PacketProcessors.Logistics
+{
+    [RegisterPacketProcessor]
+    public class ILSgStationPoolSyncRequestProcessor : IPacketProcessor<ILSRequestgStationPoolSync>
+    {
+        private PlayerManager playerManager;
+        private const int maxShipsPerILS = 10;
```
Comment (Baldie-dev) [inline NebulaHost/PacketProcessors/Logistics/ILSgStationPoolSyncRequestProcessor.cs:17]: Can this be unified with `const int ILSMaxShipCount = 10;` which is in the `NebulaClient.PacketProcessors.Logistics.ILSgStationPoolSyncProcessor`?

For example having this in both cases: `ILSShipManager.ILSMaxShipCount`

File: NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs [hunk 26]
```diff
@@ -0,0 +1,74 @@
...
+                }
+                if(stationComponent != null)
+                {
+                    StationStore[] storage = stationComponent.storage;
+                    int[] itemId = new int[storage.Length];
+                    int[] itemCountMax = new int[storage.Length];
+                    int[] itemCount = new int[storage.Length];
+                    int[] localLogic = new int[storage.Length];
+                    int[] remoteLogic = new int[storage.Length];
+                    int[] remoteOrder = new int[storage.Length];
```
Comment (Baldie-dev) [inline NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs:42]: I think there is space for future improvement, to have these arrays pre-allocated and just zero it here. Or maybe it does not need to be cleaned, since all values are being replaced right? So maybe just reusing the same arrays, so we would prevent allocation on every packet.

File: NebulaModel/Packets/Logistics/ILSShipDock.cs [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+            this.shipPos = new Double3[shipPos.Length];
+            this.shipRot = new Float4[shipRot.Length];
+            this.shipPPosTemp = new Double3[shipPPosTemp.Length];
+            this.shipPRotTemp = new Float4[shipPRotTemp.Length];
+
+            this.stationGId = stationGId;
+            this.shipDockPos = new Float3(shipDockPos);
+            this.shipDockRot = new Float4(shipDockRot);
+
+            Array.Copy(shipOtherGId, this.shipOtherGId, shipOtherGId.Length);
```
Comment (Baldie-dev) [inline NebulaModel/Packets/Logistics/ILSShipDock.cs:36]: Why the copy is needed here? Would there be an issue if the whole array will be passed?
```cs
 this.shipOtherGId = shipOtherGId;
 ```
 Since this packet is being sent and the array is not modified, right?

File: NebulaModel/Packets/Logistics/ILSgStationPoolSync.cs [hunk 28]
```diff
@@ -0,0 +1,100 @@
...
+                                    int[] shipIndex,
+                                    Double3[] shipPos,
+                                    Float4[] shipRot,
+                                    Float3[] shipVel,
+                                    float[] shipSpeed,
+                                    Float3[] shipAngularVel,
+                                    Double3[] shipPPosTemp,
+                                    Float4[] shipPRotTemp)
+        {
+            this.stationGId = (int[])stationGId.Clone();
```
Comment (Baldie-dev) [inline NebulaModel/Packets/Logistics/ILSgStationPoolSync.cs:70]: Do we need to clone it here? Will, there be any issue with just:
```cs
this.stationGId = stationGId;
```

File: NebulaModel/Packets/Logistics/StationUIInitialSync.cs [hunk 29]
```diff
@@ -0,0 +1,70 @@
...
+                                    long energy,
+                                    long energyPerTick,
+                                    int[] itemId,
+                                    int[] itemCountMax,
+                                    int[] itemCount,
+                                    int[] localLogic,
+                                    int[] remoteLogic,
+                                    int[] remoteOrder)
+        {
+            this.itemId = new int[itemId.Length];
```
Comment (Baldie-dev) [inline NebulaModel/Packets/Logistics/StationUIInitialSync.cs:43]: Same with before, is it possible to just assign the reference?
```cs
this.itemId = itemId;
```

File: NebulaPatcher/NebulaPlugin.cs [hunk 30]
```diff
@@ -4,6 +4,7 @@
 using NebulaPatcher.Logger;
 using NebulaPatcher.MonoBehaviours;
 using System;
+using System.IO;
```
Comment (Baldie-dev) [inline NebulaPatcher/NebulaPlugin.cs:7]: I think this dependency is not used, right?

**Review by Baldie-dev — APPROVED** [review]
